### ISSUE TITLE
Accessibility WCAG issues

### ISSUE THREAD

**Issue by spektred** [mapbox/mapbox-gl-js#13337]
**mapbox-gl-js version 2.5.1**:

Hello, my employer is working with an ADA agency and they made us aware of 8 accessibility issues pertaining to our mapbox installation and it appears that these are global issues which reside outside our access to resolve. Perhaps many mapbox users have the same ADA issues that we uncovered. 

So I was wondering, should I submit each violation as a separate issue here on github? Or should I simply post all 8 issues here?  Or is there a developer here who would be willing to just knock out each of them if I send over the spreadsheet?  Some of them are quite easy (i.e. add a "title" element as a child of the "svg" element )

Thanks!

**Comment by underoot** [comment]
Hey, @spektred. Thank you for your report!

I think it would be good if you share a link to the spreadsheet or the list with short descriptions right here, and after that, we can decide what to do next.

**Comment by spektred** [comment]
Hi, thanks for responding to my thread @underoot !  I know ADA issues can be a challenge to resolve.

After reviewing all of the issues again, I spotted a few more than I was able to resolve on my own, but here are the remaining 4 items which appear outside of our control. 

### 1.

> [Issue]
> There are interactive controls that cannot be operated with the keyboard alone. Examples include:
> - "Reset bearing to north" map control

This is regarding the maps 'Display zoom and rotation' control. Without using a mouse, you can use the "tab" key on your keyboard to access the zoom-in/zoom-out buttons (which are inside the map) and you can click on them using the "enter" key, but when you tab down to the rotation control, there doesn't appear to be any way to use this element without a mouse. Clicking enter does nothing. Perhaps it could be modified to respond to the arrow keys or something? 

Here's an demo in the mapbox documentation. 
https://docs.mapbox.com/mapbox-gl-js/example/navigation/

Our ADA agency continues:

> [Recommendation]
> Ensure all interactive functionality is operable with the keyboard. The best way to accomplish this is by using appropriate native controls, which come with keyboard functionality built in.
> 
> If using a native control is not possible, the control must have tabindex="0" and have JavaScript event listeners that allow the control to be activated when Enter is pressed.


### 2.

> [Issue]
> There are interactive controls that cannot be navigated to with the keyboard alone. Examples include:
> - Tooltip controls from the map


So we have a map with numerous markers/pins on our website and when you click on the markers a tooltip/popup will appear for each of them. Similar to this demo. https://docs.mapbox.com/mapbox-gl-js/example/set-popup/

However, in that example, you can navigate to the marker using the "tab" key and open the tooltip using the enter key, so its quite ADA compliant. Ours isn't doing that. 

Similar to this example here, this map has markers, but in this case the tab key cannot reach them. Why?
https://docs.mapbox.com/mapbox-gl-js/example/custom-marker-icons/

The first example is ADA compliant but the second one isn't, but ideally all maps should be.

> [Recommendation]
> Ensure all interactive functionality is operable with the keyboard. The best way to accomplish this is by using appropriate native controls, which come with keyboard functionality built in.
> 
> If using a native control is not possible, the control must have tabindex="0" and have JavaScript event listeners that allow the control to be activated when Enter is pressed.

### 3.

> [Issue]
> There are link elements with an invalid ARIA `role` attribute value of 'listitem'. Examples include:
> - (c) Mapbox
> - (c) OpenStreetMap
> - Improve this map

At the bottom right corner of every map, there are some Mapbox copyright links. For some reason these hyperlinks contain role="listitem". 

My AI client is telling me that the `role="listitem"` attribute should be applied to the `<li>` elements, not the `<a>` elements. You might need to confirm why they did this, or if it was just an error.

It also seems completely unnecessary to even have `<ul role="list">` and `<li role="listitem">` because you are telling assistive technologies what these elements are, but they can clearly see it already due to the HTML list elements.  It'd be different if you had those roles on `<div>` elements. So I think all these "roles" can be completely removed?

> [Recommendation]
> Ensure ARIA roles, states, and properties are valid. 


###  4.

> [Issue]
> There are user interface components with a contrast ratio below 3.00:1.
> 
> Foreground color: #EDAA00
> Background color: #E6E9C8
> Contrast ratio: 1.6:1

I'm not sure if you have the ability to fix this one, but it has to do with insufficient color contrast, which can impact people who are color blind. Certain parts of text in the map are hard to read. 

This example is specifically referring to the Gulf of Mexico. But there may be others like it as well in other oceans.
For example, go to this map, and zoom/scroll on over to the Gulf of Mexico. 
https://docs.mapbox.com/mapbox-gl-js/example/mapbox-gl-geocoder-outside-the-map/

Notice the blue words of "Gulf of Mexico" right on top of the blue-colored ocean. It would be very difficult for someone with poor eyesight to see that. If I had the ability to do so, I would change those words to black (#000000) and any other instance of blue text on top of a blue ocean. 

> [Recommendation]
> Ensure active user interface components have sufficient contrast. The required minimum contrast ratio is 3.00:1.



Those are all of the issues we have for now. Thanks again for any help you can provide us on this. 

And if you have any questions about any of these don't hesitate to ask! I can reach out to them and get detailed answers.

**Comment by underoot** [comment]
@spektred Thank you for the detailed and valuable description of the issues. We will take a look into it later when we will have time. Thank you again for your contribution!

**Comment by RikKeistoer** [comment]
As an addition to this, i also get a warning when i open or close a new `new mapboxgl.Popup()`.

```js
new mapboxgl.Popup({
    closeButton: true,
    closeOnClick: false,
    maxWidth: '400px',
    focusAfterOpen: true,
})
    .addTo(map)
    .setLngLat(coordinates)
    .setHTML('Some HTML here');
```

> Blocked aria-hidden on an element because its descendant retained focus. The focus must not be hidden from assistive technology users. Avoid using aria-hidden on a focused element or its ancestor. Consider using the inert attribute instead, which will also prevent focus. For more details, see the aria-hidden section of the WAI-ARIA specification at https://w3c.github.io/aria/#aria-hidden.
Element with focus: button
Ancestor with aria-hidden:  <button class=​"mapboxgl-popup-close-button" type=​"button" aria-label=​"Close popup" aria-hidden=​"true">​×​</button>​

Maybe it's my lack of mapbox knowledge, but i could not find a way to fix this console log

**Comment by devinle** [comment]
Thanks for opening this thread. I have an additional comment, granted it may be user error on my part. But flagging nonetheless.

When running an AXE scan on my mapbox component, I'm seeing this error:

**1. Required ARIA attributes must be provided.**
_aria-required-attr_
Ensure elements with ARIA roles have all required ARIA attributes

Further digging reveals a `<label>` element being added to the `<mapbox-search-listbox>` component.

In the following snippet, you can see the empty label `id="mbx0e7619e5-Label"`. Additionally, and don't quote me on this, but I don't believe `role="label"` is valid markup.

For what it's worth, adding to this post. Happy to provide more information, or move this comment elsewhere if required.

Thank you!

```
<mapbox-search-listbox data-seed="mbx0e7619e5">
  <div class="mbx0e7619e5--MapboxSearch">
    <div class="mbx0e7619e5--Label" role="label" aria-live="polite" aria-atomic="true" id="mbx0e7619e5-Label">
    </div>
    <div class="mbx0e7619e5--Results" aria-hidden="true" style="left: 201.5px; top: 1744px;">
      <div class="mbx0e7619e5--ResultsList" role="listbox" id="mbx0e7619e5-ResultsList" aria-labelledby="mbx0e7619e5-Label">
      </div>
      <div class="mbx0e7619e5--ResultsAttribution" aria-hidden="true">
        <a href="https://www.mapbox.com/search-service" target="_blank" tabindex="-1">
          Powered by Mapbox
        </a>
      </div>
    </div>
  </div>
</mapbox-search-listbox>
```

**Comment by underoot** [comment]
Hey, @spektred. I finally reviewed your list of issues and have questions, comments

> 1. <...> Clicking enter does nothing. Perhaps it could be modified to respond to the arrow keys or something?

I am unable to reproduce the issue myself. I've checked the example page https://docs.mapbox.com/mapbox-gl-js/example/navigation/, managed to rotate the map without the mouse, zoom in and out with controls, and also by pressing "Enter" managed to activate the control "Reset bearing to North", which resets bearing. See the video with a demonstration.

https://github.com/user-attachments/assets/6e35df3d-4b8d-45a5-a565-bfc49ea08241

Maybe it's a problem with some specific setup?

> 2. Similar to this example here, this map has markers, but in this case, the tab key cannot reach them. Why?

The markers in this example, indeed, cannot be accessed through the Tab key, because these markers were created with `<div>`s, and it can be fixed by adding `tabIndex` to them. I'm in the process of updating this example. Thank you!

> 3. At the bottom right corner of every map, there are some Mapbox copyright links. For some reason these hyperlinks contain role="listitem".

Could you share the specific example, I've checked couple of ours styles and seems like they don't have such an attribute.

> 4. <...> I would change those words to black (#000000) and any other instance of blue text on top of a blue ocean.

I've passed your feedback to our design team. Meanwhile, you can customize your styles, hosted in Mapbox with https://console.mapbox.com/studio/

**Comment by underoot** [comment]
> As an addition to this, i also get a warning when i open or close a new new mapboxgl.Popup().

Hey, @RikKeistoer! I think that this warning and the related a11y issue were fixed in this release https://github.com/mapbox/mapbox-gl-js/releases/tag/v3.10.0

**Comment by underoot** [comment]
Hey, @devinle! Thank you for the report. I've passed the description to the team, who are working on the `@mapbox/search-js-web` package.

**Comment by spektred** [comment]
Hello @underoot  !

Thanks for getting back to us regarding these issues. 

1. When you have the entire map selected it seems that the arrow and shift+arrow keys work fine, but when you tab down to the compass control, the arrow keys no longer appear to work?  What key did you press to make the map move when the compass was selected?

2. Great!  This is huge.

3. Ah, this issue can be disregarded. It looks like our previous ADA contractors had inserted `role="list"` and `role="listitem"` on the hyperlinks in the map because perhaps they were `<div>` elements previously, since you are now properly using `<ul>` and `<li>` those role attributes can be removed on our end. We injected them into the map. Thanks! 

4. I had no idea we had access to a mapbox console. Thanks for this information!  

Cheers

**Comment by underoot** [comment]
> What key did you press to make the map move when the compass was selected?

@spektred, I've pressed the "Enter" key. I think it's okay that the arrow keys don't work when the button is focused. The only purpose of this button is to move the bearing of the map back to north, and it seems logical that you can control the bearing of the map with the keyboard only when the map is focused.

**Comment by spektred** [comment]
@underoot 
Understood. I see that the spacebar key will make the compass control reset its position when spacebar is pressed. Also, since Google Maps doesn't even have a compass control on their desktop version of the map (only mobile) I think we can let this slide. 

I think this thread can be closed, unless anyone else here has some issues that are still standing.  Thanks again. 

Cheers

**Comment by underoot** [comment]
I'm closing this ticket then. Thank you all, @spektred, @RikKeistoer, @devinle for the feedback! 

Feel free to open a new issue if something new will be rised. Meanwhile, I will keep you updated regarding progress with the documentation update (point 2 in the first list of issues) and regarding the search box when something comes up.

**Comment by underoot** [comment]
Hey, @spektred! Fix for documentation was just released https://docs.mapbox.com/mapbox-gl-js/example/custom-marker-icons/

**Comment by spektred** [comment]
Fantastic. Is this fix pushed out to all mapbox installations, or does a new updated script need to be installed for it to be operational?

**Comment by spektred** [comment]
For example, we are pointing to:

`https://api.mapbox.com/mapbox-gl-js/v2.5.1/mapbox-gl.css`

`https://api.mapbox.com/mapbox-gl-js/plugins/mapbox-gl-geocoder/v4.7.2/mapbox-gl-geocoder.css`

and 

`https://api.mapbox.com/mapbox-gl-js/plugins/mapbox-gl-geocoder/v4.7.2/mapbox-gl-geocoder.min.js`

**Comment by underoot** [comment]
Hey, @spektred. This fix correctly handles the behaviour of `div` markers with images of this particular example. You can use it for the reference regardless version of GL JS libraries that you have
